From d531adea9a0205893e6028b09434b2a8d753aedc Mon Sep 17 00:00:00 2001
From: Chris Adams <chris.adams@jollamobile.com>
Date: Thu, 14 Nov 2013 14:40:49 +1000
Subject: [PATCH] [libaccounts-qt] Check id before attempting to load account

This commit ensures that the Accounts::Manager::account(id) function
will check the id before attempting to load the account.

It also ensures that change signals are only emitted for accounts
with a valid id.  The id may become invalidated if the account is
removed between when the change occurs and when the (queued) signal
is handled by the private class.
---
 Accounts/manager.cpp |   21 ++++++++++++++++-----
 1 file changed, 16 insertions(+), 5 deletions(-)

diff --git a/Accounts/manager.cpp b/Accounts/manager.cpp
index 448c925..afbefbd 100644
--- a/Accounts/manager.cpp
+++ b/Accounts/manager.cpp
@@ -147,22 +147,26 @@ void Manager::Private::init(Manager *q, AgManager *manager)
 
 void Manager::Private::on_account_created(Manager *self, AgAccountId id)
 {
-    Q_EMIT self->accountCreated(id);
+    if (id != 0)
+        Q_EMIT self->accountCreated(id);
 }
 
 void Manager::Private::on_account_deleted(Manager *self, AgAccountId id)
 {
-    Q_EMIT self->accountRemoved(id);
+    if (id != 0)
+        Q_EMIT self->accountRemoved(id);
 }
 
 void Manager::Private::on_account_updated(Manager *self, AgAccountId id)
 {
-    Q_EMIT self->accountUpdated(id);
+    if (id != 0)
+        Q_EMIT self->accountUpdated(id);
 }
 
 void Manager::Private::on_enabled_event(Manager *self, AgAccountId id)
 {
-    Q_EMIT self->enabledEvent(id);
+    if (id != 0)
+        Q_EMIT self->enabledEvent(id);
 }
 
 /*!
@@ -246,7 +250,14 @@ Manager::~Manager()
 Account *Manager::account(const AccountId &id) const
 {
     GError *error = NULL;
-    AgAccount *account = ag_manager_load_account(d->m_manager, id, &error);
+    AgAccount *account = NULL;
+
+    if (id == 0) {
+        qWarning() << Q_FUNC_INFO << "id is null, account has been removed or invalidated";
+        return NULL;
+    }
+
+    account = ag_manager_load_account(d->m_manager, id, &error);
 
     if (account != NULL) {
         Q_ASSERT(error == NULL);
-- 
1.7.10.4

